### DISCUSSION TITLE
How to designate SDT probe arguments for eBPF and bpftrace?

### DISCUSSION THREAD

**Discussion by aapoalas** [bpftrace/bpftrace#3615] [Q&A]
Hello

I've been adapting a DTrace USDT probe generating Rust crate to generate SystemTap's v3 USDT probe definitions (See [here](https://github.com/oxidecomputer/usdt/compare/master...aapoalas:usdt:feat/linux-stap-support) for the code.). As the main source material I've been using a C/C++ header only [USDT library](https://github.com/libbpf/usdt/blob/46c09f9440257639afc9f8cbf9dcb51f01dd9929/usdt.h#L356-L385)'s code.

I've gotten to the point where the USDTs are properly defined, bpftrace, perf, and other Linux tracing tools detect them and can attach to them, and the probes themselves work in code. What is not working is arguments: For a NT_STAPSDT probe reported by `readelf` like this:

```text
  stapsdt              0x00000037	NT_STAPSDT (SystemTap probe descriptors)
    Provider: nova
    Name: start_builtin_call
    Location: 0x00000000005c7bcd, Base: 0x0000000000d2c376, Semaphore: 0x0000000000f4a0f4
    Arguments: 8@%rdi
```

running a bpftrace script with `{ printf("%u \n", $#); }` prints only 0s. Trying to print `str(arg0)` (the first parameter is a C string in this case) will just make the script exit because there is no arguments, of course. I don't quite understand what I'm doing wrong: As far as I understand, the arguments format should be correct but obviously I'm doing _something_ wrong somewhere. An additional point of interest is that I saw [in a past issue](https://github.com/bpftrace/bpftrace/issues/33#issuecomment-419717523) that there at least was talk of a custom arguments format or STAP format extension, but I have been unable to find any documentation on this.

Is there somewhere I could read up on the format that bpftrace expects? Possibly also on how bpftrace's expectations line up with the eBPF landscape in general; does bpftrace have a custom arguments format that is different from others or has the eBPF world chosen a single format to stick with?

A bonus question I had: How expensive are probe semaphores? As I understand it, in DTrace the semaphore's isenabled data is in the instruction stream directly, and DTrace will rewrite that on the fly so that the program just runs as normal but the instructions change. Hence, an isenabled probe is essentially free. In the STAP world the isenabled data is held in the rw data section and we have to dereference that pointer to read the data, which I'd expect to cause some level of dcache trashing, making semaphores somewhat non-trivial in their performance cost. Is my understanding correct?

**Comment by aapoalas** [comment]
After changing various things while changing nothing, the arguments passing now started to magically work.

That being said, I am still quite interested in hearing if there is any documentation on the arguments format that eBPF programs expect beyond the STAP format [defined here in all its intricate glory](https://sourceware.org/systemtap/wiki/UserSpaceProbeImplementation).

And also the performance cost of semaphores is of interest to me.

**Comment by fbs** [comment]
>  running a bpftrace script with { printf("%u \n", $#); } prints only 0s. 
`$#` prints the amount of  arguments given to bpftrace, it has nothing to do with the probe. https://github.com/bpftrace/bpftrace/blob/master/man/adoc/bpftrace.adoc#positional-parameters has an example

The codegen code can be found here: https://github.com/bpftrace/bpftrace/blob/master/src/ast/irbuilderbpf.cpp

**Reply by aapoalas** [reply]
Thanks, I managed to realise this mistake of mine as well. I think my mistakes that made the arguments not work were:
1. Trying to use a DTrace's isenabled-like separate probe for isenabled data which had a semaphore but no arguments, guarding the actual probe with arguments but no semaphore. I didn't realise that in STAP there is only one probe and the semaphore performs the duties of DTrace's `isenabled` probe. As a result my bpftrace was likely only hitting the semaphore'd, argument-less probe and the main probe was probably never being reached.
2. Not being careful enough with the parameter docs: I did realise `$1` etc would be entirely wrong, but somehow still ended up trying to use `$#` for its alluring promise of actually telling me if bpftrace was successfully parsing my arguments format.

**Comment by fbs** [comment]
https://github.com/bpftrace/bpftrace/issues/3601 might be interesting
